key-rotator/manifest/manifest.go

```diff
@@ -56,7 +61,22 @@ type UpdateKeysConfig struct {
 	PacketEncryptionKeyCSRFQDN  string  // the FQDN to specify for packet encryption key CSRs
 }
 
+func (cfg UpdateKeysConfig) Validate() error {
+	if cfg.BatchSigningKey.IsEmpty() {
+		return errors.New("batch signing key has no key versions")
```

Comment:
Nit: These errors are being referenced by the test. I'd make them a var so you can just directly grab them instead of having to do string matching. Also gives flexibility in case you need to do something when you see these errors when this actually runs __for realz__

Reply:
I'd argue that using string-matching in tests is the right path, for now.

These error conditions aren't part of the public API of the package. That is, we don't expose these errors in a way that would allow other packages to discriminate between the different error cases/types/kinds -- errors are currently just a human-readable descriptive string. We don't expose error kinds because we don't need to, and I don't implement functionality that I don't need for YAGNI reasons. (If things changed and we did start exposing error types as a public part of the package, i.e. because some other package needed to discriminate between the different ways that `Validate`/`UpdateKeys` can fail, it would be absolutely correct to test against the public API.)

While I don't want to expand the package's public API, I could define package-private errors and check against them via `errors.Is`. When the error string is a constant string, as on this line, this would be easy enough (`var batchSigningKeyHasNoVersions = errors.New("...")`). But several errors are not constant strings (e.g. L140, or any other error coming from `fmt.Errorf`). To test against these in a type-safe way, I'd need to implement a new error type for each error condition as a private implementation detail, then use `errors.Is` against those new types. But ... testing against a private error type is IMO no better[1] than testing against a private error string, and it is more code to implement.

WDYT?

[1] Modulo, I suppose, a private error type could reword the human-readable string returned from `Error()` without needing to update the relevant test. I'm OK with that tradeoff.

Reply:
I think this is fine, honestly this test actually ensures that if we're breaking changes on error messages we REALLY need to be aware of it. LGTM